File: backend/app/schemas/azure_devops.py
from pydantic import BaseModel


class AzureDevOpsHours(BaseModel):
    original: float = 0
    remaining: float = 0
    completed: float = 0


class AzureDevOpsLinkedItem(BaseModel):
    id: int
    type: str | None = None
    title: str | None = None
    state: str | None = None


class AzureDevOpsWorkItemOut(BaseModel):
    id: int
    type: str | None = None
    title: str | None = None
    state: str | None = None
    area_path: str | None = None
    iteration_path: str | None = None
    assigned_to: str | None = None
    hours: AzureDevOpsHours
    parent: AzureDevOpsLinkedItem | None = None
    epic: AzureDevOpsLinkedItem | None = None


class AzureDevOpsTotalsOut(BaseModel):
    total: int
    with_epic: int
    without_epic: int
    by_state: dict[str, int]
    hours: AzureDevOpsHours


class AzureDevOpsDiagnosticItemOut(BaseModel):
    id: int
    title: str | None = None
    state: str | None = None
    type: str | None = None
    parent: AzureDevOpsLinkedItem | None = None
    epic: AzureDevOpsLinkedItem | None = None


class AzureDevOpsDiagnosticByUserOut(BaseModel):
    user: str
    count: int
    items: list[AzureDevOpsDiagnosticItemOut]


class AzureDevOpsDiagnosticGroupOut(BaseModel):
    total: int
    users: list[AzureDevOpsDiagnosticByUserOut]


class AzureDevOpsDiagnosticsOut(BaseModel):
    pbi_without_task: AzureDevOpsDiagnosticGroupOut
    tasks_without_hours: AzureDevOpsDiagnosticGroupOut


class AzureDevOpsSnapshotOut(BaseModel):
    items: list[AzureDevOpsWorkItemOut]
    totals: AzureDevOpsTotalsOut
    diagnostics: AzureDevOpsDiagnosticsOut

File: frontend/src/api/azureDevops.ts
import api from './client';

export interface AzureDevOpsLinkedItem {
  id: number;
  type?: string | null;
  title?: string | null;
  state?: string | null;
}

export interface AzureDevOpsWorkItem {
  id: number;
  type?: string | null;
  title?: string | null;
  state?: string | null;
  area_path?: string | null;
  iteration_path?: string | null;
  assigned_to?: string | null;
  hours: {
    original: number;
    remaining: number;
    completed: number;
  };
  parent?: AzureDevOpsLinkedItem | null;
  epic?: AzureDevOpsLinkedItem | null;
}

export interface AzureDevOpsSnapshot {
  items: AzureDevOpsWorkItem[];
  totals: {
    total: number;
    with_epic: number;
    without_epic: number;
    by_state: Record<string, number>;
    hours: {
      original: number;
      remaining: number;
      completed: number;
    };
  };
  diagnostics: {
    pbi_without_task: {
      total: number;
      users: Array<{
        user: string;
        count: number;
        items: Array<{
          id: number;
          title?: string | null;
          state?: string | null;
          type?: string | null;
          parent?: AzureDevOpsLinkedItem | null;
          epic?: AzureDevOpsLinkedItem | null;
        }>;
      }>;
    };
    tasks_without_hours: {
      total: number;
      users: Array<{
        user: string;
        count: number;
        items: Array<{
          id: number;
          title?: string | null;
          state?: string | null;
          type?: string | null;
          parent?: AzureDevOpsLinkedItem | null;
          epic?: AzureDevOpsLinkedItem | null;
        }>;
      }>;
    };
  };
}

export interface AzureDevOpsSnapshotParams {
  project?: string;
  team?: string;
  area_path?: string;
  iteration_path?: string;
  top?: number;
}

export const getAzureDevOpsSnapshot = async (connectorId: number, params: AzureDevOpsSnapshotParams) => {
  const { data } = await api.get<AzureDevOpsSnapshot>(`/azure-devops/connectors/${connectorId}/snapshot`, { params });
  return data;
};
